tsl/src/compression/recompression.c

```diff
@@ -0,0 +1,1460 @@
+											&next_tuple))
+			{
+				/* decompress affected batch */
+				decompress_in_tuplestore(decompressor, next_tuple, tuplestorecxt->tuplestore);
+			}
+			total_batches_to_decompress--;
+		}
+		tuplesort_performsort(tuplestorecxt->tuplestore);
+		recompress_segment(tuplestorecxt->tuplestore, decompressor->out_rel, row_compressor);
+		cleanup_tuplestorecxt(tuplestorecxt);
+		compressed_row_count++;
+		affectedsegments->total_affected_compressed_rows--;
+	}
+	ExecDropSingleTupleTableSlot(compressed_slot);
+}
+
+static void
+do_full_recompression(RowDecompressor *decompressor, RowCompressor *row_compressor,
+					  const CompressTuplesortstateCxt *uncompressed_tuplestorecxt)
+{
+	Relation idxrel;
+	IndexScanDesc index_scan;
+	TupleTableSlot *compressed_slot = NULL;
+
+	idxrel = index_open(row_compressor->index_oid, AccessShareLock);
+	index_scan = index_beginscan(decompressor->in_rel, idxrel, GetTransactionSnapshot(), 0, 0);
+	index_rescan(index_scan, NULL, 0, NULL, 0);
+	compressed_slot = table_slot_create(decompressor->in_rel, NULL);
+
+	while (index_getnext_slot(index_scan, ForwardScanDirection, compressed_slot))
+	{
+		slot_getallattrs(compressed_slot);
+		HeapTuple tuple = heap_form_tuple(compressed_slot->tts_tupleDescriptor,
+										  compressed_slot->tts_values,
+										  compressed_slot->tts_isnull);
+		tuple->t_self = compressed_slot->tts_tid;
+		decompress_in_tuplestore(decompressor, tuple, uncompressed_tuplestorecxt->tuplestore);
+		heap_freetuple(tuple);
+	}
+	ExecDropSingleTupleTableSlot(compressed_slot);
+	index_endscan(index_scan);
+	index_close(idxrel, AccessShareLock);
+
+	tuplesort_performsort(uncompressed_tuplestorecxt->tuplestore);
+	recompress_segment(uncompressed_tuplestorecxt->tuplestore,
+					   decompressor->out_rel,
+					   row_compressor);
+}
+
+static void
+recompress_new_segments(const RowDecompressor *decompressor, RowCompressor *row_compressor,
+						AffectedSegmentsCxt *affectedsegments,
+						const CompressTuplesortstateCxt *uncompressed_tuplestorecxt,
+						const ColumnCompressionInfo **keys, const int n_keys)
+{
+	TupleTableSlot *uncompressed_slot;
+	int count = 0;
+	uncompressed_slot = MakeTupleTableSlot(decompressor->out_desc, &TTSOpsMinimalTuple);
+	while (affectedsegments->total_new_compressed_segments)
+	{
+		CompressTuplesortstateCxt *tuplestorecxt;
+		tuplestorecxt = initialize_tuplestore(decompressor->out_rel, keys, n_keys);
+
+		while (affectedsegments->num_non_matching_compressedrows[count] &&
+			   tuplesort_gettupleslot(uncompressed_tuplestorecxt->tuplestore,
+									  true /*=forward*/,
+									  false /*=copy*/,
+									  uncompressed_slot,
+									  NULL /*=abbrev*/))
+		{
+			tuplesort_puttupleslot(tuplestorecxt->tuplestore, uncompressed_slot);
+			affectedsegments->num_non_matching_compressedrows[count]--;
+		}
+		tuplesort_performsort(tuplestorecxt->tuplestore);
+		recompress_segment(tuplestorecxt->tuplestore, decompressor->out_rel, row_compressor);
```

Comment:
Let's make our custom copy of `row_compressor_process_ordered_slot` here? How this would help us:

1) We won't need the `CompressionType` parameter. In general I think it's not a good interface, passing a flag to a function that says "do this" or "do that". Instead we should have different functions. But here, the purpose of the `CompressionType` flag is just to pass it down to the row compressor, so that it can assign the proper sequence numbers. Instead of that, the recompression code should take control of sequence numbers. We know exactly which rows we are going to compress and what sequence numbers the batches should have. So the old interface of "row compressor" that just works row-by-row and assigns sequence numbers itself is not suitable here.

2) `row_compressor_process_ordered_slot` does some extra work inside like `row_compressor_update_group`, but we don't need it here, we know that the rows we are recompressing belong to the same segments. So this is another way in which the old row compressor interface is not appropriate here, and leads to extra work.